Revert to old tacos unlock

kneeyo1 opened the pull request:
This is the older version (no matrix, need to remove)  to prevent tacos unlock from missing slices that were touched in the PR but later removed. 

this is a successful run https://github.com/getsentry/tacos-gha.demo/pull/2423#issuecomment-2016110168

sideop committed:
reverting back to old version of tacos unlock

Commit:
auto-commit: GHA deps, for test

sideop committed:
add back terragrunt-slices

sideop committed:
this might look cleaner

sideop committed:
revert a bit more, no more fan in summary

sideop committed:
chmod

sideop committed:
triyng this script

sideop committed:
correctly run tf-lock-release

sideop committed:
this gets all slices and not just changed files

sideop committed:
fully remove the matrix workflow and do ALL slices

sideop committed:
remove unused var

sideop committed:
needs the matrix for setup

sideop committed:
lets get this bread

kneeyo1 commented on .github/workflows/tacos_unlock.yml:
```diff
@@ -70,13 +69,11 @@ jobs:
       id-token: write
-
     env:
       TF_ROOT_MODULE: ${{matrix.tf-root-module}}
```
this is still needed because the setup step breaks otherwise. 

https://github.com/getsentry/tacos-gha.demo/actions/runs/8425696855/job/23079804120

if we really wanted to remove this, it would require also editing/reverting `./tacos-gha/.github/actions/setup` which every other workflow is using, so just leaving it like this is better.

bukzor replied:
No I think leaving things in this state is causing the system to get into an inconsistent state.

bukzor commented on .github/workflows/tacos_unlock.yml:
```diff
@@ -70,13 +69,11 @@ jobs:
       matrix:
         tf-root-module:
           ${{ fromJSON(needs.determine-tf-root-modules.outputs.slices) }}
```
the matrix is still here :(

bukzor reviewed (changes requested):
same comment as ever: there should be no matrix

sideop committed:
remove the matrix stuff, add conditional for set-workload-identity-provider

sideop committed:
clean code is happy code

sideop committed:
add some checks for tf root module

sideop committed:
bash tricks

sideop committed:
workload-identity-provider gets passed quite a bit of stuff

sideop committed:
trying this one more time, it should work

sideop committed:
new setup

sideop committed:
update pr comment

sideop committed:
chmod

sideop committed:
more chmod

sideop committed:
i think this is supposed to be a pipe

sideop committed:
added terragrunt and terraform setup

sideop committed:
clean up ocid_provider call

sideop committed:
json dumps needs list

sideop committed:
trying this out to make json happy

sideop committed:
properly set the output

sideop committed:
this should unlock all slices, matrix style

sideop committed:
if statement might be broken. fixed

sideop committed:
add debugging step

sideop committed:
im missing a pipe

sideop committed:
this might be why

sideop committed:
stealing all the setup tasks

sideop committed:
custom setup probably..

sideop committed:
# This is a combination of 7 commits.
# This is the 1st commit message:

debugging

# This is the commit message #2:

debugging

# This is the commit message #3:

debugging

# This is the commit message #4:

proper json encoding

# This is the commit message #5:

convert list to json string

# This is the commit message #6:

trying this out instead of all slices

# This is the commit message #7:

pipe in all the slices first

sideop committed:
direct add in the editted files using slices.py

bukzor commented on lib/tf_lock/release.py:
```diff
@@ -184,8 +187,13 @@ def main() -> None:
+        successes = 0
         for path in paths:
-            tf_lock_release(path, env=environ.copy())
+            if tf_lock_release(path, env=environ.copy()) is None:
```
You're dropping the UserError on the ground if/when it's returned.

kneeyo1 replied:
It gets chalked up as a failure, I think the errors are written out and that's handled seperately, we can just have it return 1 or 0 and just add that?

bukzor commented on .github/workflows/tacos_unlock.yml:
```diff
@@ -89,78 +101,35 @@ jobs:
-      - name: Save matrix result
-        # we need to show any errors to end-users
+          # release all tfstate locks currently held 
+          quietly \
+          echo "$SLICES" | jq -r '.[]' | ./tacos-gha/lib/ci/unlock
```
This says `quietly echo`, meaning the middle third of `echo`'s output will be elided if it succeeds with more than 30 lines of output. I don't believe this is what you intended.

bukzor commented on lib/ci/list-terraformers:
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -euo pipefail
+
+python3 -m lib.tacos.terraformers |
+jq -R |
+jq -cs |
+gha-set-output terraformers |
+# prettify
+jq . \
+;
```
Needs an indent:

```suggestion
python3 -m lib.tacos.terraformers |
  jq -R |
  jq -cs |
  gha-set-output terraformers |
  # prettify
  jq . \
;
```

bukzor commented on lib/tacos/terraformers.py:
```diff
@@ -0,0 +1,92 @@
+        )
+
+
+def lines_to_paths(lines: Iterable[str]) -> Generator[OSPath]:
+    for line in lines:
+        line = line.strip()
+        if not line or line.startswith("#"):
+            continue
+        yield OSPath(line)
```
This is copy-pasted from somewhere. Import it instead?

bukzor commented on lib/tacos/terraformers.py:
```diff
@@ -0,0 +1,92 @@
+        yield OSPath(line)
+
+
+def convert_terraform_result(
+    result: TerraformerResult,
+) -> Dict[str, Union[str, List[str]]]:
+    """Convert TerraformerResult to a JSON-serializable dictionary"""
+    return {
+        "GETSENTRY_SAC_OIDC": result.GETSENTRY_SAC_OIDC,
+        "SUDO_GCP_SERVICE_ACCOUNT": result.SUDO_GCP_SERVICE_ACCOUNT,
+        # Convert each OSPath in the set to a string, then convert the set to a list
+        "slices": [str(path) for path in result.slices],
+    }
```
If you change the definition of `slices` to `tuple[str]` then `dataclasses.asdict` will do this for you.

kneeyo1 replied:
wouldn't slices then just be a tuple that can only hold one string?

bukzor commented on lib/tf_lock/release.py:
```diff
@@ -121,32 +121,35 @@ def get_lock_info(root_module: Path) -> Tuple[bool, dict[str, str]]:
-            raise UserError(code=error.returncode)
+            return UserError(code=error.returncode)
 
         info(f"tf-lock-release: success: {root_module}({lock_user})")
+        return None
```
Yea this always prints something, succeed or fail, so I can't understand the blank summary in your test PR comment: https://github.com/getsentry/tacos-gha.demo/pull/2423#issuecomment-2035482202

kneeyo1 replied:
Wait, doesn't it only print if it success? if it fails, it will return the UserError code, unless I'm missing something

kneeyo1 replied:
I think the blank lines are from here https://github.com/getsentry/tacos-gha/blob/835d145cde99c8707fd0d9723e58c27c7fd4f5ed/lib/ci/unlock#L10-L13

sideop committed:
simplified tf_lock_release

remove lines_to_paths

update success

update failure

sideop committed:
correct details for failure

correct details for failure

need the heading detail tag

sideop committed:
refactored to create basic setup derived from setup

sideop committed:
removing one more step

add to both

sideop committed:
add gcp auth

sideop committed:
add the inputs

sideop committed:
give it the SA as input

sideop committed:
need to pass in the ssh-private-key

sideop committed:
trying this first

trying gcp auth again

spaces

spaces

debug

gcp auth in the right place

sideop committed:
write out debug msg for UserError

sideop committed:
tf lock acquire exit is not consistent, tthis fixes it

bukzor committed:
make tf-lock use terragrunt-noninteractive

prettier

sideop committed:
fixed working dir

sideop committed:
Merge branch 'main' into fix/unlock-all-on-release

kneeyo1 committed:
Merge branch 'main' into fix/unlock-all-on-release

kneeyo1 committed:
Update TODO.md
